partitura/musicanalysis/note_array_to_score.py

```diff
@@ -0,0 +1,358 @@
+        note_array = create_divs_from_beats(note_array)
+    elif all([x in dtypes for x in ["onset_beat", "duration_beat", "pitch"]]):
+        anacrusis_mask[note_array["onset_beat"] < 0] = True
+        note_array = create_divs_from_beats(note_array)
+    elif all([x in dtypes for x in ["onset_div", "pitch", "duration_div"]]):
+        pass
+    else:
+        raise AttributeError("The note array does not include the necessary fields.")
+
+    if "voice" in dtypes:
+        estimate_voice_info = False
+        part_voice_list = note_array["voice"]
+    else:
+        estimate_voice_info = True
+        part_voice_list = np.full(len(note_array), np.inf)
+
+    if not all(x in dtypes for x in ['step', 'alter', 'octave']):
+        warnings.warn("pitch spelling")
+        spelling_global = analysis.estimate_spelling(note_array)
+        note_array = rfn.merge_arrays((note_array, spelling_global), flatten=True)
+
+    if estimate_voice_info:
+        warnings.warn("voice estimation", stacklevel=2)
+        # TODO: deal with zero duration notes in note_array.
+        # Zero duration notes are currently deleted
+        estimated_voices = analysis.estimate_voices(note_array)
+        assert len(part_voice_list) == len(estimated_voices)
+        for i, (part_voice, voice_est) in enumerate(zip(part_voice_list, estimated_voices)):
+            # Not sure if correct.
+            if part_voice != np.inf:
+                estimated_voices[i] = part_voice
+        note_array = rfn.append_fields(note_array, "voice", np.array(estimated_voices, dtype=int))
+
+    if estimate_key or ('ks_fifths' not in dtypes and 'ks_mode' not in dtypes):
+        warnings.warn("key estimation", stacklevel=2)
+        k_name = analysis.estimate_key(note_array)
+        global_key_sigs = [[0, k_name]]
+    else:
+        if key_sigs is None and "ts_beats" in dtypes:
+            global_key_sigs = [[0, fifths_mode_to_key_name(note_array[0]["ks_fifths"], note_array[0]["ks_mode"])]]
+            for n in note_array:
+                if n["ts_beats"] != global_key_sigs[-1][1] or n["ts_beat_type"] != global_key_sigs[-1][2]:
+                    global_key_sigs.append([n["onset_div"], fifths_mode_to_key_name(n["ks_fifths"], n["ks_mode"])])
+            else:
+                global_key_sigs = key_sigs
+    global_key_sigs = np.array(global_key_sigs)
+    # for convenience, we add the end times for each time signature
+    ks_end_times = np.r_[global_key_sigs[1:, 0], np.max(note_array["onset_div"]+note_array["duration_div"])]
+    global_key_sigs = np.column_stack((global_key_sigs, ks_end_times))
+
+    # compute quarter divs
+    if divs is None:
+        # find first note with nonzero duration (in case score starts with grace_note).
+        for idx, dur in enumerate(note_array["duration_beat"]):
+            if dur != 0:
+                break
+        divs = int((note_array[idx]["duration_div"] / note_array[idx]["duration_beat"])*(note_array[idx]["ts_beat_type"]/4))
```

Comment:
if a note_array comes from a part set to `musical_beats`, this computation will give a wrong result

Reply:
Ok, that is slightly problematic, would you have a proposal to fix this?

Reply:
I just document it.